File: cs/Minsk/CodeAnalysis/Syntax/SyntaxTree.cs
using System.Collections.Immutable;
using Minsk.CodeAnalysis.Text;

namespace Minsk.CodeAnalysis.Syntax;

public sealed class SyntaxTree
{
    private SyntaxTree(SourceText text)
    {
        Text = text;

        var parser = new Parser(text);
        Root = parser.ParseCompilationUnit();
        Diagnostics = parser.Diagnostics;
    }

    public SourceText Text { get; }
    public CompilationUnitSyntax Root { get; }
    public ImmutableArray<Diagnostic> Diagnostics { get; }

    public static SyntaxTree Parse(string text)
    {
        var sourceText = SourceText.From(text);
        return Parse(sourceText);
    }

    public static SyntaxTree Parse(SourceText text)
    {
        return new SyntaxTree(text);
    }

    public static IEnumerable<SyntaxToken> ParseTokens(string text)
    {
        var sourceText = SourceText.From(text);
        return ParseTokens(sourceText);
    }

    public static IEnumerable<SyntaxToken> ParseTokens(SourceText text)
    {
        return new Lexer(text);
    }
}

File: kt/Minsk/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/SyntaxTree.kt
package dev.phytolizer.minsk.analysis.syntax

import dev.phytolizer.minsk.analysis.Diagnostic
import dev.phytolizer.minsk.analysis.text.SourceText

class SyntaxTree private constructor(val text: SourceText) {
    val root: CompilationUnitSyntax
    val diagnostics: List<Diagnostic>

    init {
        val parser = Parser(text)
        root = parser.parseCompilationUnit()
        diagnostics = parser.diagnostics
    }

    companion object {
        fun parse(text: String): SyntaxTree {
            return parse(SourceText.from(text))
        }

        fun parse(text: SourceText): SyntaxTree {
            return SyntaxTree(text)
        }

        fun parseTokens(text: String): List<SyntaxToken> {
            return parseTokens(SourceText.from(text))
        }

        fun parseTokens(text: SourceText): List<SyntaxToken> {
            return Lexer(text).toList()
        }
    }
}
